pkg/service/backup/restore_worker_data.go

```diff
@@ -443,43 +495,104 @@ func (w *restoreWorker) newRunProgress(ctx context.Context, run *RestoreRun, tar
 		return nil, err
 	}
 
 	pr := &RestoreRunProgress{
-		ClusterID:    run.ClusterID,
-		TaskID:       run.TaskID,
-		RunID:        run.ID,
-		ManifestPath: run.ManifestPath,
-		Keyspace:     run.Keyspace,
-		Table:        run.Table,
-		Host:         h.Host,
-		AgentJobID:   jobID,
-		SSTableID:    takenIDs,
+		ClusterID:         run.ClusterID,
+		TaskID:            run.TaskID,
+		RunID:             run.ID,
+		ManifestPath:      run.ManifestPath,
+		Keyspace:          run.Keyspace,
+		Table:             run.Table,
+		Host:              h.Host,
+		AgentJobID:        jobID,
+		SSTableID:         takenIDs,
+		VersionedProgress: versionedPr,
 	}
 
 	w.insertRunProgress(ctx, pr)
+	w.metrics.UpdateFilesProgress(pr.ClusterID, pr.ManifestPath, pr.Keyspace, pr.Table, pr.VersionedProgress, 0, 0)
 
 	return pr, nil
 }
 
 // startDownload creates rclone job responsible for downloading batch of SSTables.
-func (w *restoreWorker) startDownload(ctx context.Context, host, dstDir, srcDir string, batch []string) (int64, error) {
-	jobID, err := w.Client.RcloneCopyPaths(ctx, host, dstDir, srcDir, batch)
+// Downloading of versioned files happens first in a synchronous way.
+// It returns jobID for asynchronous download of the newest versions of files
+// alongside with the size of the already downloaded versioned files.
+func (w *restoreWorker) startDownload(ctx context.Context, host, dstDir, srcDir string, batch []string) (int64, int64, error) {
+	var (
+		restoreT, _    = SnapshotTagTime(w.SnapshotTag) // nolint: errcheck
+		regularBatch   = make([]string, 0)
+		versionedBatch = make([]versionedSSTable, 0)
+		versionedPr    int64
+	)
+	// Decide which files require to be downloaded in their older version
+	for _, file := range batch {
+		if versions, ok := w.versionedFiles[file]; ok {
+			var match versionedSSTable
+			for _, v := range versions {
+				t, _ := SnapshotTagTime(v.Version) // nolint: errcheck
+				if restoreT.Before(t) {
+					match = v
+					break
+				}
+			}
+
+			if match.Version != "" {
+				versionedBatch = append(versionedBatch, match)
+				versionedPr += match.Size
+				continue
+			}
+		}
+
+		regularBatch = append(regularBatch, file)
+	}
+	// Downloading versioned files requires us to rename them (strip snapshot tag suffix)
+	// and function RcloneCopyPaths lacks this option. In order to achieve that, we move
+	// all versioned files one by one with RcloneMoveFile (which supports renaming files).
+	// The assumption is that the existence of versioned files is low and that they
+	// are rather small, so we can do it in a synchronous way.
+	// Moving files can be done in full parallel because of rclone ability to limit transfers.
+	err := parallel.Run(len(versionedBatch), parallel.NoLimit, func(i int) error {
+		file := versionedBatch[i]
+
+		dst := path.Join(dstDir, file.Name)
+		src := path.Join(srcDir, strings.Join([]string{file.Name, file.Version}, "."))
+
+		if err := w.Client.RcloneMoveFile(ctx, host, dst, src); err != nil {
```

Comment:
Why move files ? We don't want to change the bucket insides. We just want to restore using correct version of SSTable.

Reply:
It indeed should be CopyFile, my bad!